File: BRITE/Java/Model/TopDownHierModel.java
/****************************************************************************/
/*                  Copyright 2001, Trustees of Boston University.          */
/*                               All Rights Reserved.                       */
/*                                                                          */
/* Permission to use, copy, or modify this software and its documentation   */
/* for educational and research purposes only and without fee is hereby     */
/* granted, provided that this copyright notice appear on all copies and    */
/* supporting documentation.  For any other uses of this software, in       */
/* original or modified form, including but not limited to distribution in  */
/* whole or in part, specific prior permission must be obtained from Boston */
/* University.  These programs shall not be used, rewritten, or adapted as  */
/* the basis of a commercial software or hardware product without first     */
/* obtaining appropriate licenses from Boston University.  Boston University*/
/* and the author(s) make no representations about the suitability of this  */
/* software for any purpose.  It is provided "as is" without express or     */
/* implied warranty.                                                        */
/*                                                                          */
/****************************************************************************/
/*                                                                          */
/*  Author:     Alberto Medina                                              */
/*              Anukool Lakhina                                             */
/*  Title:     BRITE: Boston university Representative Topology gEnerator   */
/*  Revision:  2.0         4/02/2001                                        */
/****************************************************************************/

package Model;

import Graph.*;
import Topology.*;
import Util.*;

import java.util.ArrayList;
import java.util.Random;

public class TopDownHierModel extends Model {

    ArrayList Models;
    int edgeConn;
    int k;  
    int bwInter;
    double interMin, interMax;
  int numASNodes;

    public TopDownHierModel(ArrayList Models, int edgeConnMethod, int k, 
			    int bwInter, double interMin, double interMax,
			    int bwIntra, double intraMin,double intraMax) 
    {
	if (Models.size()<2) {
	    System.out.println("[BRITE ERROR] Top Down Hierarchical Model requires more than one level.");
	    System.exit(0);
	}
	this.Models = Models;
	this.edgeConn = edgeConnMethod;
	this.k = k;
	this.bwInter = bwInter;
	this.interMin = interMin;
	this.interMax = interMax;

    }
    
    public String toString() { 
      String modelParams = "Model ("+ ModelConstants.HI_TOPDOWN+" - TopDown)\n";
	modelParams+= ((Model)Models.get(0)).toString() + ((Model)Models.get(1)).toString();
	return modelParams;
    }
  public int getNumASNodes() { return numASNodes; }
    
    /**
     *  Node s and Node d are nodes are AS neighbors.  Here, we connect the corresponding router level
     *  graphs of s and d according to edgeConn and update our flattened topology, gFlat.  
     *
     */  
    public void ConnectNodes(Node s, Node d, Graph gFlat, int direction) {
	
	Random r = rm.EDGE_CONN();
	Graph sG =((ASNodeConf)s.getNodeConf()).getTopology().getGraph();
	Graph dG = ((ASNodeConf)d.getNodeConf()).getTopology().getGraph();

	
	Node sFlat=null;
	Node dFlat = null; 
	
	if (edgeConn == ModelConstants.TD_RANDOM) {
	    /*pick a random node from s's subgraph*/
	    int rand = (int)  (Distribution.getUniformRandom(r)*sG.getNumNodes());
	    sFlat = sG.getKthNode(rand);
	    if (sFlat == null) System.out.println("sFlat is null");
	    /*pick a random node from g's subgraph*/
	    int rand2 = (int) (Distribution.getUniformRandom(r)*dG.getNumNodes());
	    dFlat = dG.getKthNode(rand2);
	    if (dFlat == null) {
		System.out.println("dest is null");
		System.out.println("rand2 = " + rand2 +  "   and dG's numnodes = " + dG.getNumNodes() );
		dG.dumpToOutput();
		System.exit(0);
	    }
	}
	else if (edgeConn == ModelConstants.TD_SMALLEST_NONLEAF) {
	    /*recall that a leaf is a node with degree = Model.m*/
	    int m = ((Model)Models.get(0)).getM();
	    sFlat = sG.getSmallestDegreeNodeThreshold(m);
	    dFlat = dG.getSmallestDegreeNodeThreshold(m);
	}
	else if (edgeConn == ModelConstants.TD_SMALLEST) {
	    /*get smallest degree node of sG*/
	     sFlat = sG.getSmallestDegreeNode();
	    /*get smallest degree node of dG*/
	     dFlat = dG.getSmallestDegreeNode();
	}
	else if (edgeConn == ModelConstants.TD_KDEGREE) {
	    sFlat = sG.getSmallestDegreeNodeThreshold(k);
	    dFlat = dG.getSmallestDegreeNodeThreshold(k);
	
	}
	else {
	    Util.ERR("EdgeConn Method " +edgeConn + " not found/implemented yet.");
	}
	
	if (sFlat==null || dFlat==null)
	    Util.ERR("NULL NODE found when connecting topologies in TopDownHierModel.");

	
	((RouterNodeConf)sFlat.getNodeConf()).setType(ModelConstants.RT_BORDER);
	((RouterNodeConf)dFlat.getNodeConf()).setType(ModelConstants.RT_BORDER);
	
	/*connect sFlat and dFlat in gFlat graph*/
	Edge e = new Edge(sFlat, dFlat);
	e.setEdgeConf(new ASEdgeConf());
	e.setDirection(direction);
	gFlat.addEdge(e);

	/*set bandwidth between border nodes according to bwInter and interMin, interMax*/
	AssignInterBW(e);	
    }


    private void AssignInterBW(Edge e) {
	Random BWRandom = rm.BW();

	if (bwInter == ModelConstants.BW_CONSTANT)
	    e.setBW(interMin);
	else if (bwInter == ModelConstants.BW_UNIFORM)
	    e.setBW(Distribution.getUniformRandom(BWRandom));
       	else if (bwInter == ModelConstants.BW_HEAVYTAILED)
	    e.setBW(Distribution.getParetoRandom(BWRandom, interMin, interMax, 1.2));
	else if (bwInter == ModelConstants.BW_EXPONENTIAL)
	    e.setBW(Distribution.getExponentialRandom(BWRandom, interMin)); 
       	else   e.setBW(-1);
	
    }
    
    
    public void ConnectTopologies(Graph g0, Graph gFlat) {
	//if (g0.isConnected())
	//   System.out.println("g0 is connected");
	//else System.out.println("g0 isn't connected");
	ArrayList g0Edges = g0.getEdgesVector();
	int size = g0Edges.size();
	for (int i=0; i<size; ++i) {
	    Edge e = (Edge) g0Edges.get(i);
	    ConnectNodes(e.getSrc(), e.getDst(), gFlat, e.getDirection());  /*connect the routers of neighboring AS nodes according to edgeConn method*/
	}
    } 
    
    public Graph FlattenGraph(Graph g0) {
	/*initialize ArrayList to numNodes of g0 - since this is number of graphs at bottom level*/
	ArrayList VectorOfGraphs = new ArrayList(g0.getNumNodes()); 
	ArrayList g0Nodes = g0.getNodesVector();
	int size = g0Nodes.size();
	for (int i=0; i<size; ++i) {
	    Node n = (Node) g0Nodes.get(i);
	    Graph g = ((ASNodeConf)n.getNodeConf()).getTopology().getGraph();
	    VectorOfGraphs.add(g);
	}

       Graph g= new Graph(VectorOfGraphs);
       //testing:
       return g;
    }
    
    public Graph Generate() {
	/*Step 1: Generate each level of topology*/
	Util.MSG("Generating Top Level (AS) graph)");
	Graph g0 = ((Model)Models.get(0)).Generate();  //g0 is toplevel graph (AS)
	numASNodes = g0.getNumNodes();
	
	if (g0.isConnected())
	    Util.DEBUG("AS graph is connected");
	else Util.DEBUG("AS Graph ***NOT*** connected!");
	
	Node[] vec = g0.getNodesArray();
	
	/*
	if (Models.get(1) instanceof FileModel)
	    FileGraph = ((Model)Models.get(1)).generate();
	*/  
	for (int i=0; i<vec.length; ++i) {
	    Node n = vec[i];
	    if (n.getNodeConf() instanceof ASNodeConf)
		{
		    Util.MSG("Generating Router Topology # " + (i+1)+" of"+ vec.length +"...");
		    Topology t = new Topology((Model)Models.get(1)); //level1 graphs
		    
		    /*1) set this AS node's corresponding router topology to the one generated by our
		     *level one model, and 2) for all the routers in that topology, assign them asID to 
		     *be N's id
		     */
		    ((ASNodeConf)n.getNodeConf()).setTopology(t, n.getID()); 
		    
		}
	}
	
	/*Step 2: Now flatten the graph to one level*/
	Util.MSG("Flattening Graph...");
	Graph gFlat = FlattenGraph(g0);

	
	/*Step 3: Now interconnect the disconnected portions according to edgeConn*/
	Util.MSGN("Connecting router graphs according to edgeConn method...");
	ConnectTopologies(g0, gFlat);
	System.out.println("... Done.");
		
	return gFlat;
    }



  /* 
  to debug:
  =========
  
  public static void main(String args[]) {
  
  RouterWaxman rw = new RouterWaxman(1000, 1000, 1000, 1, 2, (double)0.2, (double)0.1, 1, 1, 10, 10);
  ASWaxman aw = new ASWaxman(10, 1000, 1000, 1, 2, (double)0.2,(double) 0.1, 1, 1, 10, 10);
  ArrayList models = new ArrayList();
  models.add(aw);
  models.add(rw);
  
  TopDownHierModel td = new TopDownHierModel(models, 1, 2, 1, 10, 100, 2, 500, 700);
  td.Generate();
  
  }
  */
  
    
}

File: BRITE/C++/Models/TopDownHierModel.cc
/****************************************************************************/
/*                  Copyright 2001, Trustees of Boston University.          */
/*                               All Rights Reserved.                       */
/*                                                                          */
/* Permission to use, copy, or modify this software and its documentation   */
/* for educational and research purposes only and without fee is hereby     */
/* granted, provided that this copyright notice appear on all copies and    */
/* supporting documentation.  For any other uses of this software, in       */
/* original or modified form, including but not limited to distribution in  */
/* whole or in part, specific prior permission must be obtained from Boston */
/* University.  These programs shall not be used, rewritten, or adapted as  */
/* the basis of a commercial software or hardware product without first     */
/* obtaining appropriate licenses from Boston University.  Boston University*/
/* and the author(s) make no representations about the suitability of this  */
/* software for any purpose.  It is provided "as is" without express or     */
/* implied warranty.                                                        */
/*                                                                          */
/****************************************************************************/
/*                                                                          */
/*  Author:    Alberto Medina                                               */
/*             Anukool Lakhina                                              */
/*  Title:     BRITE: Boston university Representative Topology gEnerator   */
/*  Revision:  2.0         4/02/2001                                        */
/****************************************************************************/
#pragma implementation "TopDownHierModel.h"

#include "TopDownHierModel.h"
#include "RouterWaxmanModel.h"
#include "RouterBarabasiAlbertModel-1.h"
#include "RouterBarabasiAlbertModel-2.h"
#include "ASWaxmanModel.h"
#include "ASBarabasiAlbertModel-1.h"
#include "ASBarabasiAlbertModel-2.h"
#include "ImportedFileModel.h"

TopDownHierModel::TopDownHierModel(TopDownPar* par) : models(2) {

  RouterWaxman* rt_wax_model;
  RouterBarabasiAlbert_1* rt_bar_1_model;
  RouterBarabasiAlbert_2* rt_bar_2_model;
  ASWaxman* as_wax_model;
  ASBarabasiAlbert_1* as_bar_1_model;
  ASBarabasiAlbert_2* as_bar_2_model;
  ImportedBriteTopologyModel* if_brite_model;
  ImportedGTitmTopologyModel* if_gtitm_model;  
  ImportedNLANRTopologyModel* if_nlanr_model;

  edge_conn_type = (EdgeConnType)par->GetEC();
  k = par->GetK();
  m_edges = par->GetM();
  type = TD_HIER;

  cout << "Creating TD model...\n" << flush;
  switch (((ModelPar*)(par->GetModelPar(0)))->GetModelType()) {
  case AS_WAXMAN:
    as_wax_model = new ASWaxman((ASWaxPar*)(par->GetModelPar(0)));
    models[0] = as_wax_model;
    break;

  case AS_BARABASI_1:
    as_bar_1_model = new ASBarabasiAlbert_1((ASBarabasiAlbert_1_Par*)(par->GetModelPar(0)));
    models[0] = as_bar_1_model;
    break;

  case AS_BARABASI_2:
    as_bar_2_model = new ASBarabasiAlbert_2((ASBarabasiAlbert_2_Par*)(par->GetModelPar(0)));
    models[0] = as_bar_2_model;
    break;

  case IF_AS:
    
    switch (((ImportedFilePar*)(par->GetModelPar(0)))->GetFormat()) {
    case ImportedFileModel::IF_BRITE:
      cout << "Importing brite...\n" << flush;
      if_brite_model = new ImportedBriteTopologyModel((ImportedFilePar*)(par->GetModelPar(0)));
      models[0] = if_brite_model;
      break;

    case ImportedFileModel::IF_GTITM:
    case ImportedFileModel::IF_GTITM_TS:
      cout << "Importing gtitm...\n" << flush;
      if_gtitm_model = new ImportedGTitmTopologyModel((ImportedFilePar*)(par->GetModelPar(0)));
      models[0] = if_gtitm_model;
      break;

    case ImportedFileModel::IF_NLANR:
      cout << "Importing nlanr..\n" << flush;
      if_nlanr_model = new ImportedNLANRTopologyModel((ImportedFilePar*)(par->GetModelPar(0)));
      models[0] = if_nlanr_model;
      break;
      
    case ImportedFileModel::IF_SKITTER:
      cerr << "Skitter model will be available soon...\n" << flush;
      exit(0);
      
    default:
      cerr << "TopDownHierModel(): Invalid file format for ImportedFileModel...\n" << flush;
      exit(0);
    }
    break;

  default:
    cerr << "Invalid model type for TD model...\n" << flush;
    assert(0);
  }

  switch (par->GetModelPar(1)->GetModelType()) {
  case RT_WAXMAN:
    rt_wax_model = new RouterWaxman((RouterWaxPar*)(par->GetModelPar(1)));
    models[1] = rt_wax_model;
    break;

  case RT_BARABASI_1:
    rt_bar_1_model = new RouterBarabasiAlbert_1((RouterBarabasiAlbert_1_Par*)(par->GetModelPar(1)));
    models[1] = rt_bar_1_model;
    break;

  case RT_BARABASI_2:
    rt_bar_2_model = new RouterBarabasiAlbert_2((RouterBarabasiAlbert_2_Par*)(par->GetModelPar(1)));
    models[1] = rt_bar_2_model;
    break;

  case IF_ROUTER:

    switch (((ImportedFilePar*)(par->GetModelPar(1)))->GetFormat()) {
    case ImportedFileModel::IF_BRITE:
      cout << "Importing brite...\n" << flush;
      if_brite_model = new ImportedBriteTopologyModel((ImportedFilePar*)(par->GetModelPar(1)));
      models[1] = if_brite_model;
      break;

    case ImportedFileModel::IF_GTITM:
      cout << "Importing gtitm...\n" << flush;
      if_gtitm_model = new ImportedGTitmTopologyModel((ImportedFilePar*)(par->GetModelPar(1)));
      models[1] = if_gtitm_model;
      break;

    case ImportedFileModel::IF_NLANR:
      cout << "Importing nlanr..\n" << flush;
      if_nlanr_model = new ImportedNLANRTopologyModel((ImportedFilePar*)(par->GetModelPar(1)));
      models[1] = if_nlanr_model;
      break;
      
    case ImportedFileModel::IF_SKITTER:
      cerr << "Skitter model will be available soon...\n" << flush;
      exit(0);
      
    default:
      cerr << "BottomUpHierModel(): Invalid file format for ImportedFileModel...\n" << flush;
      exit(0);
    }
    break;

  default:
    cerr << "Invalid model type for TD model...\n" << flush;
    assert(0);
  }
}


string TopDownHierModel::ToString() {
  
    //  char buf[1000];
    //  ostrstream os((char*)buf, 1000);
    //  string s;

  ostringstream os(ostringstream::out);

  os << "Model ( 5 ): " 
     << (int)edge_conn_type << " "
     <<  k  << " "
     << Scale_1 << " "
     << Scale_2 << "\n";

  os << "AS Level: ";
  switch (models[0]->GetType()) {
  case AS_WAXMAN:
    os  << ((ASWaxman*)models[0])->ToString();
    break;

  case AS_BARABASI_1:
    os  << ((ASBarabasiAlbert_1*)models[0])->ToString();
    break;

  case AS_BARABASI_2:
    os  << ((ASBarabasiAlbert_2*)models[0])->ToString();
    break;

  case IF_AS:
    os  << ((ImportedFileModel*)models[0])->ToString();
    break;

  default:
    cerr << "TopDownHierModel::ToString(): Invalid AS model type...\n" << flush;
    exit(0);

  }

  os << "Router Level: ";
  switch (models[1]->GetType()) {
  case RT_WAXMAN:
    os  << ((RouterWaxman*)models[1])->ToString() << '\0';
    break;

  case RT_BARABASI_1:
    os  << ((RouterBarabasiAlbert_1*)models[1])->ToString()  << '\0';
    break;

  case RT_BARABASI_2:
    os  << ((RouterBarabasiAlbert_2*)models[1])->ToString()  << '\0';
    break;

  case IF_ROUTER:
    os  << ((ImportedFileModel*)models[1])->ToString()  << '\0';
    break;

  default:
    cerr << "TopDownHierModel::ToString(): Invalid Router model type...\n" << flush;
    exit(0);

  }

  return string(os.str());

}


Graph* TopDownHierModel::Generate() {

  cout << "Generating Top Down hierarchical topology...\n" << flush;

  /* Generate topology according to model for level 0 */
  Graph* graph = models[0]->Generate();

  /* Check for graph connectivity */
  vector<Color> color(graph->GetNumNodes());
  vector<int> pi(graph->GetNumNodes());

  for (int i = 0; i < graph->GetNumNodes(); i++) {
    color[i] = WHITE;
  }

  graph->DFS(color, pi, 0);

  int conn = 1;
  for (int i = 0; i < graph->GetNumNodes(); i++) {
    if (color[i] == WHITE) {
      conn = 0;
    }
  }
  assert(conn);

  try {

    for (int i = 0; i < graph->GetNumNodes(); i++) {
      cout << "Generating " << i + 1 << "th Router/level topology...\n" << flush;
      assert(graph->GetNodePtr(i)->GetNodeInfo()->GetNodeType() == NodeConf::AS_NODE);
      Topology* new_topology = new Topology(models[1]);
      assert(new_topology != NULL);
      assert(new_topology->IsConnected());
      ((ASNodeConf*)graph->GetNodePtr(i)->GetNodeInfo())->SetTopology(new_topology, i);   
    }
  }
  catch (bad_alloc) {
    cerr << "ToDownHier::Generate(): Cannot allocate router-level topologies...\n" << flush; 
    exit(0);
  }

  cout << "Flattening topology...\n" << flush;
  Graph* flat_graph = FlattenGraph(graph);

  cout << "Interconnecting Border routers...\n" << flush;
  InterConnectBorders(graph, flat_graph);
  
  delete graph;

  /* Returns graph pointer to flattened topology */
  return flat_graph;
  
}


Graph* TopDownHierModel::FlattenGraph(Graph* g) {

  int n = 0;
  Topology* t;
  Graph* new_graph;
  
  for (int i = 0; i < g->GetNumNodes(); i++) {
    t = ((ASNodeConf*)(g->GetNodePtr(i)->GetNodeInfo()))->GetTopology();
    n += t->GetNumNodes();
  }

  try {
    cout << "Size of Flattened topology: " << n << "\n" << flush;
    new_graph = new Graph(n);
  }
  catch (bad_alloc) {
    cerr << "TopDownHier::FlattenGraph(): Cannot allocate new graph...\n" << flush;
    exit(0);
  }

  int index = 0;
  for (int i = 0; i < g->GetNumNodes(); i++) {

    t = ((ASNodeConf*)(g->GetNodePtr(i)->GetNodeInfo()))->GetTopology();
    assert(t != NULL);

    for (int j = 0; j < t->GetNumNodes(); j++) {

      Node* node = t->GetGraph()->GetNodePtr(j);
      ((RouterNodeConf*)(node->GetNodeInfo()))->SetASId(i);
      new_graph->AddNode(node, index);
      node->SetId(index);
      index += 1;

    }
  }

  for (int i = 0; i < g->GetNumNodes(); i++) {

    assert(g->GetNodePtr(i)->GetNodeInfo()->GetNodeType() == NodeConf::AS_NODE);

    /* Grab topology associated with AS i */
    t = ((ASNodeConf*)(g->GetNodePtr(i)->GetNodeInfo()))->GetTopology();
    assert(t != NULL);

    list<Edge*>::iterator el;
    for (el = t->GetGraph()->edges.begin(); el != t->GetGraph()->edges.end(); el++) {
    
      Edge* edge = *el;
      assert(edge != NULL);
      new_graph->AddEdge(edge);
      new_graph->AddAdjListNode(edge->GetSrc()->GetId(), edge->GetDst()->GetId());
      new_graph->AddAdjListNode(edge->GetDst()->GetId(), edge->GetSrc()->GetId());
      
    }
  }
  
  return new_graph;

}


void TopDownHierModel::InterConnectBorders(Graph* g, Graph* flat_g) {

  int ASFrom, ASTo, num_nodes_i, n1, n2;
  Node *Src, *Dst;
  Edge* edge;
  vector<int> positions(g->GetNumNodes());
  RandomVariable U(s_edgeconn);

  positions[0] = 0;
  for (int i = 1; i < g->GetNumNodes(); i++) {
      
      num_nodes_i = ((ASNodeConf*)g->GetNodePtr(i)->GetNodeInfo())->GetTopology()->GetNumNodes();
      positions[i] = positions[i - 1] + num_nodes_i;
      
  }

  list<Edge*>::iterator el;
  for (el = g->edges.begin(); el != g->edges.end(); el++) {
      
    assert((*el)->GetSrc()->GetNodeInfo()->GetNodeType() == NodeConf::AS_NODE);
    assert((*el)->GetDst()->GetNodeInfo()->GetNodeType() == NodeConf::AS_NODE);
    ASFrom = (*el)->GetSrc()->GetId();
    ASTo = (*el)->GetDst()->GetId();
    double edgebw = (*el)->GetConf()->GetBW(); 

    switch (edge_conn_type) {
	
    case TD_RANDOM: 
      n1 = GetFlatRandomNode(ASFrom, g, flat_g, positions, U);
      n2 = GetFlatRandomNode(ASTo, g, flat_g, positions, U);
      break;
      
    case TD_SMALLEST:
      
      n1 = GetFlatSmallest(ASFrom, g, flat_g, positions);
      n2 = GetFlatSmallest(ASTo, g, flat_g, positions);
      break;
	
    case TD_SMALLEST_NOLEAF:
	
      n1 = GetFlatSmallestNoLeaf(ASFrom, g, flat_g, positions);
      n2 = GetFlatSmallestNoLeaf(ASTo, g, flat_g, positions);
      break;
      
    case TD_K_DEGREE:
      
      n1 = GetFlatSmallestK(ASFrom, g, flat_g, positions);
      n2 = GetFlatSmallestK(ASTo, g, flat_g, positions);
      break;
      
    default:
      cerr << "Invalid Edge Connection method...\n" << flush;
      exit(0);
    }

    Src =  flat_g->GetNodePtr(n1);
    Dst =  flat_g->GetNodePtr(n2);
    assert(Src != NULL && Dst != NULL);
    
    /* Set nodes as Border routers */
    //((RouterNodeConf*)(Src->GetNodeInfo()))->SetRouterType(RouterNodeConf::RT_BORDER);
    //((RouterNodeConf*)(Dst->GetNodeInfo()))->SetRouterType(RouterNodeConf::RT_BORDER);
    
    /* Add new edge */
    try {
      edge = new Edge(Src, Dst);
      edge->SetDirection((*el)->GetDirection());
      RouterEdgeConf* re_conf = new RouterEdgeConf(edge->Length());
      re_conf->SetBW(((ASEdgeConf*)((*el)->GetConf()))->GetBW());
      re_conf->SetEdgeType(EdgeConf::RT_EDGE);
      re_conf->SetBW(edgebw);
      edge->SetConf(re_conf);
      flat_g->AddEdge(edge);
    }
    catch (bad_alloc) {
      cerr << "TopDown::InterconnectBorders(): Cannot allocate new edege...\n" << flush;
    }
      
    /* Update adjacency lists */
    flat_g->AddAdjListNode(n1,n2);
    flat_g->AddAdjListNode(n2,n1);

  }
}


int 
TopDownHierModel::GetFlatRandomNode(int ASid, Graph* g, Graph* flat_g, vector<int>& p, RandomVariable& U) 
{
    
    int n;

    if (ASid < g->GetNumNodes() - 1) {
	n = (int)floor(U.GetValUniform(p[ASid], p[ASid + 1]));
    }else {
	n = (int)floor(U.GetValUniform(p[ASid], flat_g->GetNumNodes()));
    }

    return n;

}

int 
TopDownHierModel::GetFlatSmallest(int ASid, Graph* g, Graph* flat_g, vector<int>& p)
{
    int up;

    if (ASid < g->GetNumNodes() - 1) {
	up = p[ASid + 1];
    }else {
	up = flat_g->GetNumNodes();
    }

    int n, smallest = MAXINT;
    for (int i = p[ASid]; i < up; i++) {
	Node* node = flat_g->GetNodePtr(i);
	if (node->GetOutDegree() < smallest) {
	    n = i;
	    smallest = node->GetOutDegree();
	}
    }   

    return n;

}


int 
TopDownHierModel::GetFlatSmallestNoLeaf(int ASid, Graph* g, Graph* flat_g, vector<int>& p)
{
    int up;
    
    if (ASid < g->GetNumNodes() - 1) {
	up = p[ASid + 1];
    }else {
	up = flat_g->GetNumNodes();
    }

    int min = MAXINT;
    for(int i = 0; i < flat_g->GetNumNodes(); i++ ){
	Node* node = flat_g->GetNodePtr(i);
	if (node->GetOutDegree() < min) {
	    min = node->GetOutDegree();
	}
    }
	
    int n, smallest = MAXINT;
    for (int i = p[ASid]; i < up; i++) {
      Node* node = flat_g->GetNodePtr(i);
      if ((node->GetOutDegree() < smallest) && 
	  (node->GetOutDegree() > min)) {
	n = i;
	smallest = node->GetOutDegree();
      }
    }   
    
    return n;
    
}

int 
TopDownHierModel::GetFlatSmallestK(int ASid, Graph* g, Graph* flat_g, vector<int>& p)
{
    int up;
    
    if (ASid < g->GetNumNodes() - 1) {
      up = p[ASid + 1];
    }else {
      up = flat_g->GetNumNodes();
    }

    int n, smallest = MAXINT;
    for (int i = p[ASid]; i < up; i++) {
      Node* node = flat_g->GetNodePtr(i);
      if ((node->GetOutDegree() < smallest) && 
	  (node->GetOutDegree() > k)) {
	n = i;
	smallest = node->GetOutDegree();
      }
    }   
    
    return n;
}

